terrain_simplification_ros/src/terrain_simplification_ros.cpp

```diff
@@ -0,0 +1,179 @@
+  ros_server_pub_     = ros_nh_.advertiseService("/terrain_simplification/pub", &TerrainSimplificationRos::pub, this);
+  ros_server_read_    = ros_nh_.advertiseService("/terrain_simplification/read", &TerrainSimplificationRos::read, this);
+  ros_server_get_val_ = ros_nh_.advertiseService("/terrain_simplification/get_val", &TerrainSimplificationRos::getValueAtPosition, this);
+
+  success = true;
+}
+
+bool TerrainSimplificationRos::run(
+    std_srvs::Empty::Request& request,
+    std_srvs::Empty::Response& response) {
+  setRun(true);
+  return true;
+}
+
+bool TerrainSimplificationRos::stop(
+    std_srvs::Empty::Request& request,
+    std_srvs::Empty::Response& response) {
+  setRun(false);
+  return true;
+}
+
+bool TerrainSimplificationRos::pub(
+    std_srvs::Empty::Request& request,
+    std_srvs::Empty::Response& response) {
+  if (isReady()) {
+    pubSimplifiedGridMap();
+  } else {
+    if (!isReceived()) {
+      ROS_WARN_STREAM("The elevation map was not received. "
+                      "It either must be published on topic = " << topic_elevation_map_ <<
+                      " or manually provided via setGridMap().");
+    } else {
+      ROS_WARN_STREAM("TerrainSimplification is not running. "
+                      "It must be started via the /run service call. ");
+    }
+  }
+  return true;
+}
+
+bool TerrainSimplificationRos::read(
+    std_srvs::Empty::Request& request,
+    std_srvs::Empty::Response& response) {
+  bool success = readParameters();
+  return success;
+}
+
+bool TerrainSimplificationRos::getValueAtPosition(
+    terrain_simplification_ros::GetValueAtPosition::Request &request,
+    terrain_simplification_ros::GetValueAtPosition::Response &response) {
+  Eigen::Vector2d p(request.x, request.y);
+  bool is_inside = false;
+  double val = TerrainSimplification::getValueAtPosition(request.layer, p, is_inside);
+
+  if (is_inside) {
+    if (request.layer == "simplified") {
+      response.val = val + h_nominal_;
+    } else {
+      response.val = val;
+    }
+  } else {
+    ROS_ERROR_STREAM("The provided position (" << p.transpose() << ") "
+                     "is outside the map; "
+                     "hence, the corresponding value cannot be obtained.");
+  }
+  return is_inside;
+}
+
+double TerrainSimplificationRos::getHeight(
+    const Eigen::Vector2d& location,
+    bool& is_inside,
+    const bool& apply_h_offset) {
+  double h = TerrainSimplification::getValueAtPosition("simplified", location, is_inside);
+  if (apply_h_offset) h += h_nominal_;
+  return  h;
+}
+
+double TerrainSimplificationRos::getTraversability(
+    const Eigen::Vector2d& location,
+    bool& is_inside) {
+  TerrainSimplification::getValueAtPosition("traversability", location, is_inside);
+}
+
+bool
+TerrainSimplificationRos::readParameters() {
+  if (!ros_nh_.getParam("/terrain_simplification_ros_node/topic_elevation_map", topic_elevation_map_)){
```

Comment:
getParam("topic_elevation_map") should work ?

Reply:
I use `ros::NodeHandle node_handle("~");` and the parameter is currently defined in the main `.launch` file. What method do you recommend? Overall, It is not the most useful as you cannot change the topic without restarting the node.

Reply:
I think you don't need to specify the name of the node

> ros_nh_.getParam("topic_elevation_map", topic_elevation_map_)